flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java

```diff
@@ -0,0 +1,429 @@
+		if (inputStatus == InputStatus.END_OF_INPUT) {
+			endSorting();
+			return addNextToQueue(new HeadElement(idx), output);
+		}
+
+		return inputStatus;
+	}
+
+	@Nonnull
+	@SuppressWarnings({"unchecked"})
+	private InputStatus emitNextAfterSorting(DataOutput<IN> output) throws Exception {
+		if (commonContext.isFinishedEmitting(idx)) {
+			return InputStatus.END_OF_INPUT;
+		} else if (commonContext.allSorted()) {
+			HeadElement head = commonContext.getQueueOfHeads().peek();
+			if (head != null && head.inputIndex == idx) {
+				HeadElement headElement = commonContext.getQueueOfHeads().poll();
+				output.emitRecord((StreamRecord<IN>) headElement.streamElement.f1);
+				return addNextToQueue(headElement, output);
+			} else {
+				return InputStatus.NOTHING_AVAILABLE;
+			}
+		} else {
+			return InputStatus.NOTHING_AVAILABLE;
+		}
+	}
+
+	private void endSorting() throws Exception {
+		sorter.finishReading();
+		commonContext.setFinishedSorting(idx);
+		sortedInput = sorter.getIterator();
+		if (commonContext.allSorted()) {
+			commonContext.getAllFinished().getUnavailableToResetAvailable().complete(null);
+		}
+	}
+
+	@Nonnull
+	private InputStatus addNextToQueue(HeadElement reuse, DataOutput<IN> output) throws Exception {
+		Tuple2<byte[], StreamRecord<IN>> next = sortedInput.next();
+		if (next != null) {
+			reuse.streamElement = getAsObject(next);
+			commonContext.getQueueOfHeads().add(reuse);
+		} else {
+			commonContext.setFinishedEmitting(idx);
+			if (seenWatermark > Long.MIN_VALUE) {
+				output.emitWatermark(new Watermark(seenWatermark));
+			}
+			return InputStatus.END_OF_INPUT;
+		}
+
+		if (commonContext.allSorted()) {
+			HeadElement headElement = commonContext.getQueueOfHeads().peek();
+			if (headElement != null) {
+				if (headElement.inputIndex == idx) {
+					return InputStatus.MORE_AVAILABLE;
+				}
+			}
+		}
+
+		return InputStatus.NOTHING_AVAILABLE;
+	}
+
+	@SuppressWarnings({"rawtypes", "unchecked"})
+	private Tuple2<byte[], StreamRecord<Object>> getAsObject(Tuple2<byte[], StreamRecord<IN>> next) {
+		return (Tuple2<byte[], StreamRecord<Object>>) (Tuple2) next;
+	}
+
+	@Override
+	public boolean isApproximatelyAvailable() {
+		if (sortedInput != null) {
+			return isHeadAvailable();
+		} else {
+			return StreamTaskInput.super.isApproximatelyAvailable();
+		}
+	}
+
+	@Override
+	public boolean isAvailable() {
+		if (sortedInput != null) {
+			return isHeadAvailable();
+		} else {
+			return StreamTaskInput.super.isAvailable();
+		}
+	}
+
+	private boolean isHeadAvailable() {
+		if (!commonContext.allSorted()) {
+			return false;
+		}
+		HeadElement headElement = commonContext.getQueueOfHeads().peek();
+		return headElement != null && headElement.inputIndex == idx;
+	}
+
+	@Override
+	public CompletableFuture<?> getAvailableFuture() {
+		if (sortedInput != null) {
+			return commonContext.getAllFinished().getAvailableFuture();
+		} else {
+			return wrappedInput.getAvailableFuture();
+		}
+	}
```

Comment:
Why are you implementing `isAvailable` and `isApproximatelyAvailable` independently? There should be no performance benefit of the default implementation `getAvailableFuture() == AVAILABLE`, and I would be afraid that it can create opportunities for those methods to drift out of sync.

So far in other places we were able to keep relatively simple contract, with controlling the availability via marking `AvailabilityHelper` available or not. Maybe let's sync off line on Monday?